chaosgcp/iam/controls/policy.py

```diff
@@ -0,0 +1,304 @@
+        ) from e
+
+
+def manage_time_bound_temporary_iam_roles(
+    project_id: str,
+    members: list[str],
+    roles: list[str],
+    roles_iam_condition: dict[str, str],
+    manage_type: str,
+    iam_propogation_sleep_time_in_minutes: int = 2,
+) -> None:
+    """Manages temporary, time-bound IAM roles for the specified project and members.
+
+    This function either grants or revokes the specified IAM roles to/from the
+    provided members, with an expiration time determined by the given condition.
+
+    Args:
+        project_id: The ID of the project where the roles will be managed.
+        members: A list of member email addresses or IDs to grant/revoke roles to.
+        roles: A list of IAM role names to grant/revoke.
+        roles_iam_condition: A dictionary representing the IAM condition that
+            enforces the expiration time. It should have the following keys:
+            - "title": A descriptive title for the condition.
+            - "description": A brief explanation of the condition's purpose.
+            - "expression": The Common Expression Language (CEL) expression that
+                defines the condition's logic.
+        manage_type: The type of operation to perform: "add" to grant roles,
+            "remove" to revoke roles.
+        iam_propogation_sleep_time_in_minutes: The number of minutes to wait for
+            role changes to propagate before proceeding further (default: 2).
+
+    Raises:
+        TypeError: If any of the arguments are not of the expected type.
+        ValueError: If `manage_type` is not "add" or "remove".
+        # (Add other potential exceptions based on your IAM implementation)
+
+    Example:
+        # Grant roles with a 15-minute expiration:
+        expiration_timestamp = datetime.datetime.now() + datetime.timedelta(minutes=15)
+        condition = {
+            "title": "TemporaryAccess",
+            "description": "Grants access for 15 minutes",
+            "expression": f"request.time < timestamp(\"{expiration_timestamp.strftime('%Y-%m-%dT%H:%M:%SZ')}\")",
+        }
+        manage_time_bound_temporary_iam_roles(
+            project_id="my-project",
+            members=["user1@example.com"],
+            roles=["roles/editor"],
+            roles_iam_condition=condition,
+            manage_type="add",
+        )
+    """
+    # Type checks for arguments
+    if not isinstance(project_id, str):
+        raise TypeError("project_id must be a string")
+    if not isinstance(members, list):
+        raise TypeError("members must be a list of strings")
+    if not isinstance(roles, list):
+        raise TypeError("roles must be a list of strings")
+    if not isinstance(roles_iam_condition, dict):
+        raise TypeError("roles_iam_condition must be a dictionary")
+    if not isinstance(manage_type, str):
+        raise TypeError("manage_type must be a string")
+    if not isinstance(iam_propogation_sleep_time_in_minutes, int):
```

Comment:
in theory this isn't really needed as `chaos validate` should care for this check. worth verifying if that's the case.

Reply:
removed